### PR TITLE
add device agnostic tests & add helpful assertion for device agnostic tests

### PR DESCRIPTION
**Pull request by bgerstle**
Our app doesn't have a window when run as a test host, which results in `keyWindow` being `nil`.  This could also happen if your test target is a "Logic Test", and not an "Application Test."

```
$ ls -1 WikipediaUnitTests/ReferenceImages_64/NSAttributedString_WMFModifyTests
test0ReferenceOutput_iPhone_Simulator8_4_0x0@2x.png
testChangeGreenBackgroundColorsToBlue_iPhone_Simulator8_4_0x0@2x.png
testChangeReduceLineSpacing_iPhone_Simulator8_4_0x0@2x.png
testRemoveAllStrikeThroughs_iPhone_Simulator8_4_0x0@2x.png
testRemovingLink_iPhone_Simulator8_4_0x0@2x.png
```

I've added an assertion so others can get clearer feedback about what's going wrong and how to fix it.  FBSnapshotTestCase could arguably create a `keyWindow` implicitly, but I figured it's better to let users set up their test host properly instead of implicitly relying on the library to do it for them.

### CODE REVIEW

**Comment by bgerstle** [comment]
looks like I need to download iOS 9 sim and re-record snapshots. downloading now... :hourglass_flowing_sand:

File: FBSnapshotTestCase/FBSnapshotTestCasePlatform.m [hunk 1]
```diff
@@ -34,7 +34,7 @@ BOOL FBSnapshotTestCaseIs64Bit(void)
 NSString *FBDeviceAgnosticNormalizedFileName(NSString *fileName)
 {
   UIDevice *device = [UIDevice currentDevice];
-  CGSize screenSize = [[UIApplication sharedApplication] keyWindow].bounds.size;
```
Comment (Grubas7) [inline FBSnapshotTestCase/FBSnapshotTestCasePlatform.m]: I'm not sure if this change is OK.
UIWindow != UIScreen -> UIWindow is unique for application (especially useful for split view), and UIScreen is for whole device.

File: FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+- (void)setUp {
+    [super setUp];
+    self.deviceAgnostic = YES;
+    self.recordMode = NO;
+}
+
+- (void)tearDown {
+    // Put teardown code here. This method is called after the invocation of each test method in the class.
+    [super tearDown];
+}
```
Comment (Grubas7) [inline FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m]: Empty method. Should be delated.

File: FBSnapshotTestCase/FBSnapshotTestCasePlatform.m (hunk 1 above)
Comment (bgerstle) [inline FBSnapshotTestCase/FBSnapshotTestCasePlatform.m]: That's true, but I figured using the screen would:

- Be equivalent to the window in most cases
- Work when the library is integrated as a framework (where `keyWindow` does not)

Maybe a more comprehensive way to solve this would be to generate the suffix based on the view being tested, which is probably what you really want anyway.  This makes even more sense when you consider that snapshots are generated based on the bounds of the specified view, even creating and "keying" a window if necessary.[0]

0: https://github.com/facebook/ios-snapshot-test-case/blob/master/FBSnapshotTestCase/Categories/UIImage+Snapshot.m#L46

File: FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m (hunk 2 above)
Comment (bgerstle) [inline FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m]: :ok_hand:

File: FBSnapshotTestCase/FBSnapshotTestCasePlatform.m (hunk 1 above)
Comment (nscoding) [inline FBSnapshotTestCase/FBSnapshotTestCasePlatform.m]: hmm, we could check if the `keyWindow` is `nil` and if it is, create a window with the `[[UIScreen mainScreen] bounds]`

How does this sound?

Comment (bgerstle) [inline FBSnapshotTestCase/FBSnapshotTestCasePlatform.m]: Actually, just had a realization: the problem doesn't have anything to do with frameworks.  Our app [short-circuits itself when being run as the test host](https://github.com/wikimedia/wikipedia-ios/blob/master/Wikipedia/Code/main.m#L10), which means no `AppDelegate`, which means no `keyWindow`.

So, this is more about handling our specific edge case, which means that a window should be made the `keyWindow` before suffixes are generated.  At which point, I'm wondering if we should be using a "dummy" app delegate (which just has a dummy window), or FBSnapshotTestCase should install one for me (e.g. if you were running a Logic Tests target which didn't even have a host).

Comment (Grubas7) [inline FBSnapshotTestCase/FBSnapshotTestCasePlatform.m]: I had similar problem:
I write some framework which is using UIApplication. Same framework doesn't have UIApplication, so I created TestApp, which hosts tests.
So check if this resolve your problem:
1. Add new Application target
2. Choose your framework test target
3. Go to `general` tab
4. Set your new Application target as `Host Application`
5. Make sure that your Application target contains framework tests in test schema

Comment (nscoding) [inline FBSnapshotTestCase/FBSnapshotTestCasePlatform.m]: nice, if that works maybe we can assert and explain how to solve this

Comment (bgerstle) [inline FBSnapshotTestCase/FBSnapshotTestCasePlatform.m]: @Grubas7 the short-circuit is intentional, and our app is already the "Host" for our tests.  This removes the unnecessary "link in both targets" step when adding new app code.

My question to you guys is, should FBSnapshotTestCase be defensive by setting the `keyWindow` itself (which it already tries to do), and if so, should it guarantee that `keyWindow

Comment (bgerstle) [inline FBSnapshotTestCase/FBSnapshotTestCasePlatform.m]: My question about determines whether the solution should be _either_:

- Add an assertion and explain what users should do (set a host app or otherwise create a key window when visual tests are running)
- Create a `keyWindow` manually if necessary

**Comment by bgerstle** [comment]
@Grubas7 @nscoding I reverted my change using `UIScreen` and added an assertion instead.  This should prompt users (like me) who don't have a key window to go ahead and create one.

File: FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m [hunk 3]
```diff
@@ -0,0 +1,34 @@
+/*
```
Comment (Grubas7) [inline FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m]: Are those test still valid?

**Comment by Grubas7** [comment]
@bgerstle Cool. I wonder there should be some info about this in README. @nscoding ?

File: FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    [super setUp];
+    self.deviceAgnostic = YES;
+    self.recordMode = NO;
+}
+
+- (void)testViewSnapshot
+{
+    UIView *redView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 40, 40)];
+    redView.backgroundColor = [UIColor redColor];
+    FBSnapshotVerifyView(redView, nil);
```
Comment (nscoding) [inline FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m]: Travis is hitting the assertion that you added, why?

Comment (bgerstle) [inline FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m]: Good question, will have to investigate..

Comment (bgerstle) [inline FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m]: Because the `keyWindow` is `nil`... :confounded: Even though the app delegate is certainly [making it key and visible](https://github.com/facebook/ios-snapshot-test-case/blob/master/FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemo/FBAppDelegate.m#L22). Any ideas?

Comment (bgerstle) [inline FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m]: Also @Grubas7 this is why the tests were added

Comment (bgerstle) [inline FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m]: Should I:

- Somehow wait for the application to have a key window, maybe using `expectationWithPredicate`?
- Fall back to the application's delegate's window?

Comment (nscoding) [inline FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m]: hmm I just cloned your PR, [UIApplication sharedApplication] is nil at this point. I believe mainly because of the project doesn't have a host application.

screen shot 2015-12-16 at 11 23 43 am

Comment (nscoding) [inline FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m]: Either we fix the project to have a host application (I am leaning toward this) or we don't rely on `[UIApplication sharedApplication]` but I don't see a clean way to get the window.

Thoughts?

Comment (bgerstle) [inline FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m]: Oh wow, didn't even think to check that... When I ran the tests locally I put a break point in `FBAppDelegate` and saw it was being called.  I'll fix that straightaway.

Comment (bgerstle) [inline FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m]: @nscoding were you checking the demo project or the framework?  The demo project has a test host, which corroborates what I saw when debugging:

image

As the device agnostic tests I added are in the demo project, I don't think changing the framework test target to have a host will make a difference here.

Comment (nscoding) [inline FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m]: oh, I was look at the project itself, you are right, this is configured correctly but yet `[UIApplication sharedApplication]` is still `nil`

Comment (bgerstle) [inline FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m]: what do you suggest? poll `keyWIndow` in `FBSnapshotTestCase.setUp`? e.g.:

``` objc
- (void)setUp 
{
  // other stuff
  [self expectationWithPredicate:[NSPredicate predicateWithBlock:^ (...) {
    return evaluatedObject.keyWindow != nil;
  } ...];
}
```

Comment (nscoding) [inline FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m]: Sorry, I had to look at it a bit to figure out what is going on and there are a couple of issues:

1. The build script that I added on travis is building and testing both [iPhone 5 and iPhone 6](https://github.com/facebook/ios-snapshot-test-case/blob/master/build.sh#L25-L26), so we have to record the image for both of these devices.  

2. The [build script](https://github.com/facebook/ios-snapshot-test-case/blob/master/build.sh#L7-L11) runs the FBSnapshotControllerTests and in particular [testFailedImageWithDeviceAgnosticShouldHaveModelOnName](https://github.com/facebook/ios-snapshot-test-case/blob/master/FBSnapshotTestCaseTests/FBSnapshotControllerTests.m#L69). These tests don't have an application associated with, so we hit the assertion. Maybe we can make the window a parameter and simply write tests to make sure that FBDeviceAgnosticNormalizedFileName is working as expected.

Comment (nscoding) [inline FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m]: @bgerstle thoughts on my latest comment?

Comment (nscoding) [inline FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m]: oh I didn't see the updated commits

Comment (bgerstle) [inline FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m]: @nscoding yeah, sorry I didn't respond, but better late than never:

> The build script that I added on travis is building and testing both iPhone 5 and iPhone 6, so we have to record the image for both of these devices.

I already figured that out when Travis failed and downloaded/recorded the proper screenshots.

> testFailedImageWithDeviceAgnosticShouldHaveModelOnName. These tests don't have an application associated with, so we hit the assertion. 

Those tests are in the Demo project, which does have a host application (see above screenshot).  I _think_ the problem was that `UIImage+Snapshot` was causing issues when installing a "dummy" window when given a view without a window.  This somehow caused the `keyWindow` to return `nil` when device-agnostic tests were run—which I determined after seeing each test pass in isolation.  As a result, I've modified `UIImage+Snapshot` to use the application's `keyWindow` if the given view doesn't have a window (or if it isn't a window itself).  Let me know if you think this is acceptable.

> Maybe we can make the window a parameter and simply write tests to make sure that FBDeviceAgnosticNormalizedFileName is working as expected.

That could be useful for verifying the formatting of test names (to prevent unintended regressions and thus break tests for users), but doesn't address the problems targeted by this PR.

**Comment by bgerstle** [comment]
@Grubas7 @nscoding this should be ready to go, see my [latest comment](https://github.com/facebook/ios-snapshot-test-case/pull/137#discussion_r48477202) for a brief explanation on the approach and c9a9530 for the implementation

File: FBSnapshotTestCase/Categories/UIApplication+StrictKeyWindow.m [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+//
+//  Created by Brian Gerstle on 12/23/15.
+//  Copyright © 2015 Facebook. All rights reserved.
+//
+
+#import <FBSnapshotTestCase/UIApplication+StrictKeyWindow.h>
+
+@implementation UIApplication (StrictKeyWindow)
+
+- (UIWindow*)fb_strictKeyWindow {
```
Comment (Grubas7) [inline FBSnapshotTestCase/Categories/UIApplication+StrictKeyWindow.m]: Missing space before asterisk.

Comment (bgerstle) [inline FBSnapshotTestCase/Categories/UIApplication+StrictKeyWindow.m]: Have you guys considered using uncrustify or clang-format? 😅 While I'm at it, should I remove the comment headers (Created by Brian Gerstle etc.)?

Comment (bgerstle) [inline FBSnapshotTestCase/Categories/UIApplication+StrictKeyWindow.m]: went ahead and did it anyway

File: FBSnapshotTestCase/Categories/UIApplication+StrictKeyWindow.m [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#import <FBSnapshotTestCase/UIApplication+StrictKeyWindow.h>
+
+@implementation UIApplication (StrictKeyWindow)
+
+- (UIWindow *)fb_strictKeyWindow {
+  UIWindow *keyWindow = [[UIApplication sharedApplication] keyWindow];
```
Comment (Grubas7) [inline FBSnapshotTestCase/Categories/UIApplication+StrictKeyWindow.m]: `keyWindow` is a property, so please use dot notation.

File: FBSnapshotTestCase/Categories/UIImage+Snapshot.m [hunk 7]
```diff
@@ -57,6 +62,11 @@ + (UIImage *)fb_imageForView:(UIView *)view
 
   UIImage *snapshot = UIGraphicsGetImageFromCurrentImageContext();
   UIGraphicsEndImageContext();
+
+  if (removeFromSuperview) {
```
Comment (Grubas7) [inline FBSnapshotTestCase/Categories/UIImage+Snapshot.m]: `if ([window.subviews containObject:view])` IMO opinion is less error prone.

Comment (bgerstle) [inline FBSnapshotTestCase/Categories/UIImage+Snapshot.m]: I disagree, since there's an edge case where the condition you're testing is actually setup by the user, and therefore would be incorrect for us to tear down. The intent here is to only tear down what we ourselves setup on line 50/55.

File: FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+@implementation FBSnapshotTestCaseDeviceAgnosticTests
+
+- (void)setUp
+{
+    [super setUp];
+    self.deviceAgnostic = YES;
+    self.recordMode = NO;
+}
+
+- (void)testViewSnapshot
```
Comment (Grubas7) [inline FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m]: Test name too general.

File: FBSnapshotTestCase/Categories/UIImage+Snapshot.m (hunk 7 above)
Comment (Grubas7) [inline FBSnapshotTestCase/Categories/UIImage+Snapshot.m]: OK, you right.

File: FBSnapshotTestCaseTests/FBSnapshotControllerTests.m [hunk 9]
```diff
@@ -66,24 +66,6 @@ - (void)testCompareReferenceImageWithVeryLowToleranceShouldMatch
     XCTAssertTrue([controller compareReferenceImage:referenceImage toImage:testImage tolerance:.001 error:nil]);
 }
 
-- (void)testFailedImageWithDeviceAgnosticShouldHaveModelOnName
```
Comment (Grubas7) [inline FBSnapshotTestCaseTests/FBSnapshotControllerTests.m]: Why did you remove this test?

File: FBSnapshotTestCase/FBSnapshotTestCasePlatform.m [hunk 10]
```diff
@@ -34,7 +35,8 @@ BOOL FBSnapshotTestCaseIs64Bit(void)
...
+  UIWindow* keyWindow = [[UIApplication sharedApplication] fb_strictKeyWindow];
```
Comment (Grubas7) [inline FBSnapshotTestCase/FBSnapshotTestCasePlatform.m]: Pointer alignment ;)

**Comment by Grubas7** [comment]
Looks good :+1:

**Comment by bgerstle** [comment]
@Grubas7 thanks for the feedback! all your comments should be addressed :ok_hand:

**Comment by Grubas7** [comment]
@bgerstle could you squash your commits?

**Comment by bgerstle** [comment]
@Grubas7 squashed & amended commit message

File: FBSnapshotTestCase/Categories/UIApplication+StrictKeyWindow.m [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#import <FBSnapshotTestCase/UIApplication+StrictKeyWindow.h>
+
+@implementation UIApplication (StrictKeyWindow)
+
+- (UIWindow *)fb_strictKeyWindow {
```
Comment (nscoding) [inline FBSnapshotTestCase/Categories/UIApplication+StrictKeyWindow.m]: new line before {

File: FBSnapshotTestCase/Categories/UIApplication+StrictKeyWindow.h [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#import <UIKit/UIKit.h>
+
+@interface UIApplication (StrictKeyWindow)
+
+/**
+  @return The recdeiver's @c keyWindow. Raises an assertion if @c nil.
```
Comment (nscoding) [inline FBSnapshotTestCase/Categories/UIApplication+StrictKeyWindow.h]: recdeiver's  typo

File: FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+@implementation FBSnapshotTestCaseDeviceAgnosticTests
+
+- (void)setUp
+{
+    [super setUp];
+    self.deviceAgnostic = YES;
+    self.recordMode = NO;
+}
+
+- (void)testSuccessfulComparisonOfDeviceAgnosticReferenceImages
```
Comment (nscoding) [inline FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m]: can we also run this test (and record the image) for an iPhone 6?

**Comment by nscoding** [comment]
Thank you for working on this, it looks great, I left a couple of comments (nit picks really).

File: FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m (hunk 13 above)
Comment (bgerstle) [inline FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemoTests/FBSnapshotTestCaseDeviceAgnosticTests.m]: Interestingly, it's already run on both 5 & 6 iPhone models, but the application itself isn't filling the screen:

image

I'll see if I can fix this quickly.  Good catch that this should be verified for different screen sizes.

**Comment by bgerstle** [comment]
@nscoding addressed all of your comments, including the iPhone 6 screen size issue (and re-recorded reference image)

**Comment by nscoding** [comment]
testSuccessfulComparisonOfDeviceAgnosticReferenceImages tests are failing, it's looking for 320x568 (iPhone 5 right? but we didn't provide this one)

**Comment by bgerstle** [comment]
Ahh that's different now too, as it used to be 320x480, forgot to re-record
those

**Comment by bgerstle** [comment]
@nscoding should be good now :checkered_flag:

File: FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemo/Launch Screen.storyboard [hunk 14]
```diff
@@ -0,0 +1,50 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<document type="com.apple.InterfaceBuilder3.CocoaTouch.Storyboard.XIB" version="3.0" toolsVersion="9531" systemVersion="15C50" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" launchScreen="YES" useTraitCollections="YES" initialViewController="01J-lp-oVM">
```
Comment (nscoding) [inline FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemo/Launch Screen.storyboard]: Do we need this?

File: FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemo/Images.xcassets/LaunchImage.launchimage/Contents.json [hunk 15]
```diff
@@ -4,15 +4,134 @@
       "orientation" : "portrait",
       "idiom" : "iphone",
       "extent" : "full-screen",
+      "minimum-system-version" : "8.0",
```
Comment (nscoding) [inline FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemo/Images.xcassets/LaunchImage.launchimage/Contents.json]: I guess this file was introduced for the device size right?

File: FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemo/Launch Screen.storyboard (hunk 14 above)
Comment (bgerstle) [inline FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemo/Launch Screen.storyboard]: Yes, the launch screen fixes the issue where the app doesn't fill the
screen.

File: FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemo/Images.xcassets/LaunchImage.launchimage/Contents.json (hunk 15 above)
Comment (bgerstle) [inline FBSnapshotTestCaseDemo/FBSnapshotTestCaseDemo/Images.xcassets/LaunchImage.launchimage/Contents.json]: Yeah, one of those things that changed just by touching it. Didn't think it
was worth messing with git to revert it.
